documentation/specs/src/economics/fee-system.md

```diff
@@ -1,7 +1,378 @@
+  /// The optional unshielding tx for fee payment
+  pub unshield: Option<Tx>,
+  /// the encrypted payload
+  pub inner_tx: EncryptedTx,
+  /// sha-2 hash of the inner transaction acting as a commitment
+  /// the contents of the encrypted payload
+  pub tx_hash: Hash,
+}
+```
+
+The new `unshield` field carries an optional tx encoding for an unshielding
+`Transfer`. The unshielding operation is exempt from paying fees and doesn't
+charge gas.
+
+The proposer and the validators must also check the validity of the optional
+unshielding transfer attached to the wrapper. More specifically the correctness
+implies that:
+
+1. The unshielding provides just the right amount of funds to pay fees
+2. The actual wasm execution runs successfully
+
+The first condition can be tested statically and requires that:
+
+1. The tx encodes a `Transfer`
+2. The `shielded` field must be set to `Some`
+3. The `source` address must be the masp4. The `target` address matches that of
+   the wrapper signer
+4. The `token` match the one specified in the `Fee` struct
+5. The `amount`, added to the already available unshielded balance for that
+   token, is just enough to cover the fees, i.e. the value given by
+   $Fee.amount * GasLimit$ (to prevent leveraging this transfer for other
+   purposes)
+
+The spending key associated with this operation could be relative to any address
+as long as the signature of the transfer itself is valid.
+
+If checks 1 to 5 fail, the transaction can be safely discarded, while if the
+check fails at point 6 the transaction could be kept in mempool for future
+usage, until the expiration time of the transaction is reached.
+
+Once these controls have been performed, the block proposer should run the
+actual transfer against the current state of the application to check whether
+the transaction is valid or not: if this succeeds the transaction can be
+included in the block, otherwise it should be discarded.
+
+These same checks are done by the validators in `process_proposal`: if any of
+them fail, the entire block is rejected. The balance key must be searched in the
+local cache before the storage to ensure a correct computation in case of
+transactions involving the same addresses.
+
+### Governance proposals
+
+Governance [proposals](../base-ledger/governance.md) may carry some wasm code to
+be executed in case the proposal passed. This code is embedded into a
+`DecryptedTx` directly by the validators at block processing time and is not
+inserted into the block itself. These transactions are exempt from fees and
+don't charge gas.
+
+### Protocol transactions
+
+Protocol transactions can only be correctly crafted by validators and serve a
+role in allowing the chain to function properly. Given these, they are not
+subject to fees and do not charge gas.
+
+## Gas accounting
+
+We provide a mapping between all the whitelisted transactions and VPs to their
+cost in gas units: more specifically, the cost of a tx/VP is given by the run
```

Comment:
but this is a fixed mapping, right? the actual run-time execution speed could vary

Reply:
Yes, for the moment it will be a fixed mapping. After this first implementation is done, we'll start working on a proper run-time gas meter